### ISSUE TITLE
link with the pretrained model is dead

### ISSUE THREAD

**Issue by duynn912** [uulm-mrm/mmood3d#5]
Dear authors,

It seems the link with your model is dead. May you re-upload again?

Thank you so much!!

Sincerely.

**Comment by duynn912** [comment]
By the way, may I ask you about the MSP. To have MSP results, you use the original centerpoint not the ood model, right? Can you share where you calculate MSP in your code? I am wondering the threshold you use to know what object is ID or OOD?

**Comment by TheCodez** [comment]
Hi, thanks. I fixed the link in the README.
Yes, for the MSP results the base detector is used. When performing the evaluation of the OOD detector, all the baselines, including MSP, are also evaluated. The threshold is also printed there.

MSP code is here: https://github.com/uulm-mrm/mmood3d/blob/87c19a65b20600252e972fe9d1dae00ffeaae49c/mmood3d/models/ood_processors/ood_processors.py#L12
The OOD processors are called here: https://github.com/uulm-mrm/mmood3d/blob/87c19a65b20600252e972fe9d1dae00ffeaae49c/mmood3d/models/ood_detectors/ood_detector.py#L181

The threshold calculation is here: https://github.com/uulm-mrm/mmood3d/blob/87c19a65b20600252e972fe9d1dae00ffeaae49c/mmood3d/evaluation/functional/ood_eval.py#L101
and printed here: https://github.com/uulm-mrm/mmood3d/blob/87c19a65b20600252e972fe9d1dae00ffeaae49c/mmood3d/evaluation/metrics/ood_metric.py#L406

If you want to create your own config file you can instantiate an MSP detector like this:
```python

checkpoint = 'your base detector checkpoint path'

detector = model
detector.update(dict(
    type=OODCenterPoint,  # use version that allows better feature extraction
    pts_middle_encoder=dict(return_middle_feats=True),
))
del model

model = dict(
    type=OODDetector,
    base_detector=detector,
    base_detector_ckpt=checkpoint,
    ood_postprocessor=dict(type=MSP, threshold=0.4847),
    data_preprocessor=detector.data_preprocessor,
    train_cfg=detector.train_cfg,
    test_cfg=detector.test_cfg
)
```

**Comment by duynn912** [comment]
I first am sorry for bothering you and thank you for your nice response!

I am also working in this ood detection similarly to you. I still have some unclear issue, hope you can help me out.

1) From the paper, I understand that during ood training, the generated ood will occur. But in the ood_head.py, I see that ood_label = 10 not 11 (10 interest classes of Nuscenes Dataset + 1 OOD). Why is that?

2) For OOD generation, you generate them in each scene before training ood and save them at somewhere, or by batch during training. I mean if when you train ood model by batch, you choose 50% of them to generate after neglecting those smaller than 5 points? For example, the batch = 16, there are 8 ID objects will be chosen then generate OOD and you replace that OOD objects in the batch, at that time you have 8 ID + 8 OOD? Or you do in a different way.

3) For the logit and feature, we only have logit at BEV for you code, right? While the feature F_feat can select any feature in the network or only some at certain layers. 

Sincerely

**Comment by TheCodez** [comment]
To answer your questions:

1. The labels of the ID objects start at 0, so the last one of the 10 classes is 9. Hence I'm using 10 for OOD. In practice it should not matter what you set it to as long as it doesn't interfere with the ID classes and its consistent with the evaluator.
2. I'm generating the OOD objects online, so if in a batch there are 16 objects, and the prob is 50%, then I replace 8 objects by their OOD counterpart. The code for the augmentation is here: https://github.com/uulm-mrm/mmood3d/blob/87c19a65b20600252e972fe9d1dae00ffeaae49c/mmood3d/datasets/transforms/transforms_3d.py#L12
3. Yes the logits are extracted from the logits map in BEV. The features can be selected from multiple feature maps from the detectors "backbone", using `feature_map` in OODHead. See https://github.com/uulm-mrm/mmood3d/tree/main/mmood3d/configs/ood/ablations for the available feature maps.
